apps/emqx_s3/src/emqx_s3_profile_conf.erl

```diff
@@ -0,0 +1,391 @@
+    {noreply, State}.
+
+terminate(_Reason, #{profile_id := ProfileId}) ->
+    lists:foreach(
+        fun(PoolName) ->
+            ok = stop_http_pool(ProfileId, PoolName)
+        end,
+        emqx_s3_profile_http_pools:all(ProfileId)
+    ).
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+id(ProfileId) ->
+    {?MODULE, ProfileId}.
+
+client_config(ProfileConfig, PoolName) ->
+    HTTPOpts = maps:get(transport_options, ProfileConfig, #{}),
+    #{
+        scheme => scheme(HTTPOpts),
+        host => maps:get(host, ProfileConfig),
+        port => maps:get(port, ProfileConfig),
+        url_expire_time => maps:get(url_expire_time, ProfileConfig),
+        headers => maps:get(headers, HTTPOpts, #{}),
+        acl => maps:get(acl, ProfileConfig),
+        bucket => maps:get(bucket, ProfileConfig),
+        access_key_id => maps:get(access_key_id, ProfileConfig, undefined),
+        secret_access_key => maps:get(secret_access_key, ProfileConfig, undefined),
+        request_timeout => maps:get(request_timeout, HTTPOpts, undefined),
+        http_pool => PoolName
+    }.
+
+uploader_config(#{max_part_size := MaxPartSize, min_part_size := MinPartSize} = _ProfileConfig) ->
+    #{
+        min_part_size => MinPartSize,
+        max_part_size => MaxPartSize
+    }.
+
+scheme(#{ssl := #{enable := true}}) -> "https://";
+scheme(_TransportOpts) -> "http://".
+
+start_http_pool(ProfileId, ProfileConfig) ->
+    HttpConfig = http_config(ProfileConfig),
+    PoolName = pool_name(ProfileId),
+    case do_start_http_pool(PoolName, HttpConfig) of
+        ok ->
+            ok = emqx_s3_profile_http_pools:register(ProfileId, PoolName),
+            ok = ?tp(debug, "s3_start_http_pool", #{pool_name => PoolName, profile_id => ProfileId}),
+            {ok, PoolName};
+        {error, _} = Error ->
+            Error
+    end.
+
+update_http_pool(ProfileId, ProfileConfig, #{pool_name := OldPoolName} = State) ->
+    HttpConfig = http_config(ProfileConfig),
+    OldHttpConfig = old_http_config(State),
+    case OldHttpConfig =:= HttpConfig of
+        true ->
+            {ok, OldPoolName};
+        false ->
+            PoolName = pool_name(ProfileId),
+            case do_start_http_pool(PoolName, HttpConfig) of
+                ok ->
+                    ok = set_old_pool_outdated(State),
+                    ok = emqx_s3_profile_http_pools:register(ProfileId, PoolName),
+                    {ok, PoolName};
+                {error, _} = Error ->
+                    Error
+            end
+    end.
+
+pool_name(ProfileId) ->
+    iolist_to_binary([
+        <<"s3-http-">>,
+        ProfileId,
+        <<"-">>,
+        integer_to_binary(erlang:system_time(millisecond)),
+        <<"-">>,
+        integer_to_binary(erlang:unique_integer([positive]))
+    ]).
```

Comment:
Nit. Given that this name is used as a gproc propname and could nowadays be any term, why not just?
```suggestion
    {'s3-http-pool', ProfileId, erlang:system_time(millisecond), erlang:unique_integer([positive])}.
```

Reply:
`ehttpc` currently does not like terms as pool names  https://github.com/emqx/ehttpc/blob/main/src/ehttpc.erl#L80

Reply:
Ouch. Weird restriction. 😅 Now started to wonder if we can cheat with `['s3-http-pool', ProfileId, erlang:system_time(millisecond), erlang:unique_integer([positive])]`.